### PR TITLE
feat(prefer-in-document): add support for assigments

### PR DESCRIPTION
**Pull request by benmonro**
**What**: add support for variable assignments with prefer in document

**Why**: fixes bugs and adds more use cases. fixes: https://github.com/testing-library/eslint-plugin-jest-dom/issues/104

**How**: checking variable declarations & assignments.

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/__tests__/lib/rules/prefer-in-document.js [hunk 1]
```diff
@@ -66,13 +106,39 @@ const invalid = [
       `expect(${q}('foo')).not.toBeNull()`,
       `expect(${q}('foo')).toBeInTheDocument()`
     ),
+    invalidCase(
+      `expect(${q}('foo')) .not .toBeNull()`,
```
Comment (benmonro) [inline src/__tests__/lib/rules/prefer-in-document.js:110]: @AntonNiklasson FYI i had to change around the fixer from what you had due to this use case here.  FYI using hard coded numbers in a fixer can be problematic for stuff like spacing etc.

**Review by AntonNiklasson — CHANGES_REQUESTED** [review]
I ran this code on one of our codebases to validate, and it crashed. This case seems to be reproducing it:

```diff
diff --git a/src/__tests__/lib/rules/prefer-in-document.js b/src/__tests__/lib/rules/prefer-in-document.js
index 376dfda..bc0faf1 100644
--- a/src/__tests__/lib/rules/prefer-in-document.js
+++ b/src/__tests__/lib/rules/prefer-in-document.js
@@ -91,6 +91,12 @@ const invalid = [
       foo = screen.${q}('foo');
       expect(foo).toBeInTheDocument();`
     ),
+    invalidCase(
+      `const compressingFeedback = await screen.findByText(/Compressing video/);
+      expect(compressingFeedback).toBeDefined();`,
+      `const compressingFeedback = await screen.findByText(/Compressing video/);
+      expect(compressingFeedback).toBeInTheDocument();`
+    ),
   ]),
   // Invalid cases that applies to queryBy* and queryAllBy*
   ...queriesByVariant.query.map((q) => [
```

**Review by AntonNiklasson — CHANGES_REQUESTED** [review]
Looking good! I played around with the latest changes and managed to find another test case that currently crashes:

```
invalidCase(
  `let button;
  it("foo", async () => {
    button = await screen.findByText('click me');
    expect(button).toBeDefined();
  });`,
  `let button;
  it("foo", async () => {
    button = await screen.findByText('click me');
    expect(button).toBeInTheDocument();
  });`
)
```

I do think this "pattern" is fairly common, but it's not really feasible for this rule to support be 100% with this. I just felt that this case was pretty similar to what you already had there.

File: docs/rules/prefer-in-document.md [hunk 2]
```diff
@@ -17,17 +17,29 @@ expect(queryByText("foo")).toBeNull();
...
+
+const bar = screen.queryByText("bar");
+expect(bar).toHaveLength(0);
 ```
 
 Examples of **correct** code for this rule:
 
 ```js
 expect(screen.queryByText("foo")).toBeInTheDocument();
 expect(screen.queryByText("foo")).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (benmonro) [reply docs/rules/prefer-in-document.md:32]:
```suggestion
expect(await screen.findByText("foo")).toBeInTheDocument();
```

File: docs/rules/prefer-in-document.md [hunk 3]
```diff
@@ -17,17 +17,29 @@ expect(queryByText("foo")).toBeNull();
...
+expect(queryByText("foo")).toBeInTheDocument();
+expect(wrapper.queryAllByTestId("foo")).toBeInTheDocument();
+expect(screen.getAllByLabel("foo-bar")).toHaveLength(2);
+expect(notAQuery("foo-bar")).toHaveLength(1);
+
+const foo = screen.getAllByText("foo");
+expect(foo).toHaveLength(3);
+
+const bar = screen.queryByText("bar");
+expect(bar).not.toBeInTheDocument();
```
Comment (benmonro) [inline docs/rules/prefer-in-document.md:42]:
```suggestion
expect(bar).not.toBeDefined();

const baz = await screen.findByText("baz");
expect(baz).toBeDefined()
```

**Comment by nickmccurdy** [comment]
This isn't a code review, but FWIW this fixes the errors in https://github.com/testing-library/dom-testing-library/pull/834. Nice work!

**Comment by nickmccurdy** [comment]
We're getting various runtime failures in dom-testing-library after upgrading and linting with `--fix`. It seems like it replaces some array assertions where it shouldn't, among other things.

**Comment by benmonro** [comment]
oy.  got any examples?

**Comment by nickmccurdy** [comment]
Do you want me to just publish a branch so you can see the CI build?

**Comment by benmonro** [comment]
that would be excellent, thanks @nickmccurdy

**Comment by nickmccurdy** [comment]
https://github.com/testing-library/dom-testing-library/pull/839/checks?check_run_id=1476727442

**Comment by benmonro** [comment]
@nickmccurdy looks liek those are coming from usages of `getAllBy` that should actually be using `getBy`, no?  Would it maybe make more sense to update those?

**Pull request merged**